contributors/design-proposals/scheduling/scheduling-framework.md

```diff
@@ -1,436 +1,620 @@
+ranking of Nodes. A plugin that registers for this extension point will be
+called with the [scoring](#scoring) results from the same plugin. This is called
+once per plugin per scheduling cycle.
+
+For example, suppose a plugin `BlinkingLightScorer` ranks Nodes based on how
+many blinking lights they have.
+
+```go
+func ScoreNode(_ *v1.Pod, n *v1.Node) (int, error) {
+   return getBlinkingLightCount(n)
+}
+```
+
+However, the maximum count of blinking lights may be small compared to
+`NodeScoreMax`. To fix this, `BlinkingLightScorer` should also register for this
+extension point.
+
+```go
+func NormalizeScores(scores map[string]int) {
+   highest := 0
+   for _, score := range scores {
+      highest = max(highest, score)
+   }
+   for node, score := range scores {
+      scores[node] = scores*NodeScoreMax/highest
+   }
+}
+```
+
+If any normalize-scoring plugin returns an error, the scheduling cycle is
+aborted.
+
+**Note:** Plugins wishing to perform "pre-reserve" work should use the
+normalize-scoring extension point.
 
 ### Reserve
 
-At this point scheduler updates its cache by "reserving" a Node (partially or
-fully) for the Pod. In scheduler v1 this stage is called "assume".
-At this point, only the scheduler cache is updated to
-reflect that the Node is (partially) reserved for the Pod. The scheduling
-framework calls plugins registered at this extension points so that they get a
-chance to perform cache updates or other accounting activities. These plugins
-do not return any value (except errors).
+This is an informational extension point. Stateful plugins should use this
+extension point to be notified by the scheduler when resources on a node are
+being reserved for a given Pod. This happens before the scheduler actually binds
+the pod to the Node, and it exists to prevent race conditions while the
+scheduler waits for the bind to succeed.
 
-The actual assignment of the Node to the Pod happens during the "Bind" phase.
-That is when the API server updates the Pod object with the Node information.
+Once a pod is in the reserved state, it will either trigger
+[Un-reserve](#un-reserve) plugins (on failure) or [Post-bind](#post-bind)
+plugins (on success).
+
+*Note: This concept used to be referred to as "assume".*
 
 ### Permit
 
-Permit plugins run in a separate go routine (in parallel). Each plugin can return
-one of the three possible values: 1) "permit", 2) "deny", or 3) "wait". If all
-plugins registered at this extension point return "permit", the pod is sent to
-the next step for binding. If any of the plugins returns "deny", the pod is
-rejected and sent back to the scheduling queue. If any of the plugins returns
-"wait", the Pod is kept in reserved state until it is explicitly approved for
-binding. A plugin that returns "wait" must return a "timeout" as well. If the
-timeout expires, the pod is rejected and goes back to the scheduling queue.
+These plugins are used to prevent or delay the binding of a Pod. A permit plugin
```

Comment:
interesting, curious, what would be a use case for delaying ?

Reply:
Coscheduling is the first example that comes to mind. goal is to only schedule a set of pods if they can all run concurrently. Scheduler would process pods one by one while the coscheduling plugin holds them at this stage. When enough pods are ready, coscheduling would permit them all.